### PR TITLE
GitHub: use bitcoind v26.0 for CI

### PR DESCRIPTION
**Pull request by guggero**
Depends on https://github.com/lightningnetwork/lnd/pull/8555.

## Change Description

See if things break in the CI if we update to `bitcoind v26.0`.

### CODE REVIEW

File: .github/workflows/main.yml
Comment (calvinrzachman) [inline .github/workflows/main.yml:24]: The way we have [this job](https://github.com/lightninglabs/docker-bitcoin-core/blob/0132c0cf3974546a3f71d81c86ebe52bc5414abd/.github/workflows/build.yaml) setup, leads to images with tags like `bitcoin-core:26` and `bitcoin-core:26-alpine`. I think we may have to either drop the minor version here or re-tag and push images with a new tag that includes the minor version.

Comment (guggero) [reply .github/workflows/main.yml]: Ah, good point. Updated.

**Comment by guggero** [comment]
Okay, so the failed unit tests were due to a number of reasons:
 - `bitcoind v26.0` doesn't have support for Unix Sockets (`ipc://`) in ZMQ anymore. Broken by https://github.com/bitcoin/bitcoin/pull/22087, will be fixed by https://github.com/bitcoin/bitcoin/pull/27679
 - We had frequent TCP listening port collisions, at least when running the tests locally
 - We didn't clean up temporary directories for `bitcoind` sometimes

I added a couple of commits to address these issues.

File: scripts/install_bitcoind.sh
Comment (yyforyongyu) [inline scripts/install_bitcoind.sh:9]: I guess this assumes we gonna always build with major versions?

File: chainntnfs/test_utils.go
Comment (yyforyongyu) [inline chainntnfs/test_utils.go:176]: Feel like this can just be combined with `lntest.NewMiner` with some modifications, guess we can clean it up in future PRs.

File: lnwallet/test/test_interface.go
Comment (yyforyongyu) [inline lnwallet/test/test_interface.go:3213]: 🎉

**Review by yyforyongyu — COMMENTED** [review]
Thanks for the cleanup! No major comments, just need to merge the dependent PR first. And the unit tests need to be fixed.

File: scripts/install_bitcoind.sh
Comment (guggero) [reply scripts/install_bitcoind.sh:9]: Yeah, which I think is okay for our CI. We'll notice quite quickly if this becomes an issue.

**Review by Roasbeef — APPROVED** [review]
LGTM 🎨

File: chainntnfs/test_utils.go
Comment (guggero) [reply chainntnfs/test_utils.go:209]: We don't mention this in the user guide, only ever `tcp://`. That's why I think not many people even noticed the regression. There is no tracking issue, just a PR that aimed to fix the regression (https://github.com/bitcoin/bitcoin/pull/28020) that was closed in favor of another PR that fixed it in a more complete manner (https://github.com/bitcoin/bitcoin/pull/27679) but is farther from completion as it looks.

**Comment by yyforyongyu** [comment]
Unit tests still failing. One of the itest failures has uncleaned shutdown that still occupying a port, werid
```
 unable to create server: listen tcp 127.0.0.1:6563: bind: address already in use 
```

**Comment by guggero** [comment]
Yeah, it seems like I have to do some more digging into failures.

File: scripts/install_bitcoind.sh
Comment (yyforyongyu) [inline scripts/install_bitcoind.sh:5]: I think we should use one place to manage the version, to avoid mistakes such as 994a1426d6

Comment (ProofOfKeags) [reply scripts/install_bitcoind.sh:5]: Maybe this is a dumb question but why isn't this number 26?

**Review by ProofOfKeags — APPROVED** [review]
I pulled this down and tested it and it looks like everything passes except a wtclient test (possibly a flake?). It doesn't seem related to this as best I can tell. I think that test will pass if you rebase this.

File: scripts/install_bitcoind.sh
Comment (guggero) [reply scripts/install_bitcoind.sh:5]: This is just the fallback version in case no environment variable is set. We set the version we actually want here: https://github.com/lightningnetwork/lnd/blob/ff3e8da1756572b0dd979a39b139a969498772dd/.github/workflows/main.yml#L24

But I guess we shouldn't do that to avoid confusion...

**Comment by guggero** [comment]
@Roasbeef can you see the logs of the failed Travis build?
I only get this error message (not sure if we need to re-auth the app somehow?):

Screenshot from 2024-02-02 18-37-28

**Review by Roasbeef — CHANGES_REQUESTED** [review]
This unit test looks to be failing across a few targets now: 
```
--- FAIL: TestFilteredChainView (34.34s)
    interface_test.go:871: Testing 'bitcoind_zmq' implementation of FilteredChainView
    interface_test.go:871: Testing 'bitcoind_polling' implementation of FilteredChainView
    --- FAIL: TestFilteredChainView/bitcoind_polling:_filtered_block_ntfns (20.21s)
        interface_test.go:225: filtered block notification didn't arrive
    interface_test.go:871: Testing 'p2p_neutrino' implementation of FilteredChainView
    interface_test.go:871: Testing 'btcd_websockets' implementation of FilteredChainView
FAIL
```

**Comment by guggero** [comment]
Yeah, new failures keep popping up that I cannot easily reproduce locally. Or only with spending a lot of time on it, which I haven't found yet. Not sure if this is a priority, @saubyk?

**Comment by saubyk** [comment]
> Yeah, new failures keep popping up that I cannot easily reproduce locally. Or only with spending a lot of time on it, which I haven't found yet. Not sure if this is a priority, @saubyk?

I am not aware of any critical dependency on bitcoind v26.0 for our imminent releases, so we can keep this at a lower priority

**Comment by yyforyongyu** [comment]
Our CI is using bitcoind v23.0, and I think it's very important to test `lnd` against a newer version of `bitcoind`, there are new RPCs used but never checked in our CI, also the upcoming sweeper will also need a newer version of `bitcoind`.

Can take over if needed @guggero

**Comment by guggero** [comment]
@yyforyongyu feel free to take over! I'm sure the failures are something small, just hard to find with some rounds of debugging required.

**Comment by saubyk** [comment]
cc: @ProofOfKeags for taking this over

**Comment by ProofOfKeags** [comment]
So here's what I've found out. The success/failure of the FilteredChainView tests is highly sensitive to whether other unit tests are running as well. This makes me think that our unit test suite somehow doesn't isolate the instances of the bitcoin node it is talking to.

- If I run *just* the FilteredChainView tests, it *always* succeeds.
- If I run the tests naively (allowing test caching), it usually succeeds in the limit case. Since only passing tests cache, I suspect that repeatedly running it stabilizes the test because fewer and fewer tests are competing for the same resource each run, reducing the race-y behavior
- If I run the tests with `nocache=1` it *always* fails.
- Extending the timeout on the FilteredChainView tests does not seem to have an impact. I increased it from 20s to 30s since the test seems to reliably run in 20s. A 50% boost should be enough if it was truly a time budget issue.

Taking into consideration that this broke by upgrading from Core 23 to Core 26 is more perplexing. I'd expect race-y behavior to not depend on that.

**Comment by guggero** [comment]
Thank you @ProofOfKeags for investigating! Your insights have given me another idea what the problem might be.
The fact that running things in parallel causes the test to fail make me think the "per process unique port" generation mechanism doesn't work as expected. So I changed it to be unique per system, using a single file. Let's see how far that gets us...

**Comment by guggero** [comment]
Okay, I think I have a version that should work. Wish me luck!

**Review by ProofOfKeags — COMMENTED** [review]
So here's what I think is happening. What's going on in CI is identical afaict to what is happening for me locally.

There's a huge batch of errors related to "connection not found". Additionally, after all the tests have finished running on my machine if I inspect my Docker daemon, I find that there are some of these Postgres11 containers still running rather than being cleaned up. This leads me to believe quite strongly that there is a race condition in the resource allocator for the test instance of Postgres.

Verifying this further, I noticed that the function that is failing stating that it couldn't find the connection is the cleanup function that deallocates our resources. I bet that for some reason we are not atomically returning resources and registering deallocators.

I also noticed a possible concurrency issue wrt the filelocking behavior that will prevent intra-process concurrency from synchronizing properly.

Finally, I noticed that this is pinned to an old version of kvdb (1.2.2 + your mods).

File: go.mod
Comment (ProofOfKeags) [inline go.mod:203]: This is making it very difficult for me to try and debug this further without having to reimplement your kvdb changes on top of 1.4.5

Comment (guggero) [reply go.mod:203]: If you look at the commit, that's just what's in #8555. I didn't push all tags to my fork, that's why the Golang pseudo version looks like 1.2.2.

File: kvdb/postgres/fixture.go
Comment (guggero) [reply kvdb/postgres/fixture.go:47]: This only is being used for the `make unit tags=kvdb_postgres` tests. In the normal `make unit` and `make unit-race` this shouldn't be used. So it could explain the failures in the Postgres test, but not in the others...

File: lntest/port/port.go
Comment (guggero) [reply lntest/port/port.go:44]: Hmm... Okay. I'll need to figure out a way to do this in a way that doesn't allow other threads to access the file...

Comment (ProofOfKeags) [reply lntest/port/port.go:44]: Just also drop a global mutex in there. Mutex takes care of threads, lockfile takes care of processes.

Comment (guggero) [reply lntest/port/port.go:44]: Okay, I added in the mutex, just to be on the safe side. But reading up on the documentation, it looks like the `O_EXCL` does what we expect here:

```
O_EXCL   int = syscall.O_EXCL   // used with O_CREATE, file must not exist.
```

The `OpenFile` function (or the underlying syscall) should be thread safe and therefore we should get an error if the file already exists at that point.

```
       O_EXCL Ensure that this call creates the file: if this flag is
              specified in conjunction with O_CREAT, and pathname
              already exists, then open() fails with the error EEXIST.
```
from https://man7.org/linux/man-pages/man2/open.2.html

**Comment by guggero** [comment]
Okay, I fixed a couple of more issues:
 - Postgres: Needed to increase the number of allowed connections. Now the "connection not found" should be gone.
 - Travis: They seem to filter on the branch name and refuse to build because the word "bitcoin" appears in it. They suck anyway and GitHub now has a macOS build on M1 that we can use for our ARM (`arm64`) builds. So I replaced the Travis build with an additional GitHub run (let's see how that one fares).
 - The sporadic failures in the `routing/chainview` tests were caused by not waiting for the backend node to fully start up. There seems to be an issue with the `chain.BitcoindClient` that it can get weirdly de-synchronized if `Start()` and `NotifyBlocks()` are called while the node is syncing with the chain still (if things are very slow and blocks come in between `Start()` and `NotifyBlocks()`, weird things start to happen). Waiting one second before using a chain backend solved that issue for now.

Things look way better now IMO. There still are a couple of flakes, but I think those were all pre-existing.

**Comment by guggero** [comment]
Okay, this is as green as it will probably ever get. There are still a couple of (pre-existing!) flakes in there. But I vote to get this merged since they very likely don't have anything to do with `bitcoind`.

File: .github/workflows/main.yml
Comment (yyforyongyu) [inline .github/workflows/main.yml:332]: niiice🎉

**Review by yyforyongyu — APPROVED** [review]
LGTM, thanks for the fix🙏

**Comment by yyforyongyu** [comment]
btw can we update our merge rules to exclude the build from macOS?

File: lntest/port/port.go
Comment (ProofOfKeags) [reply lntest/port/port.go:69]: Ah I misunderstood what was happening here.

**Review by ProofOfKeags — APPROVED** [review]
🤝

**Comment by guggero** [comment]
> btw can we update our merge rules to exclude the build from macOS?

It's currently not included in the list of required tests. So merges shouldn't be blocked on it (same for Windows).

**Comment by pinheadmz** [comment]
FWIW https://github.com/bitcoin/bitcoin/pull/27679 is updated, has two ACKs and a review from an LND dev would be helpful to get a merge ;-)

**Pull request merged**
